packages/office-ui-fabric-react/src/components/Button/_base/Button.base.tsx

```diff
@@ -0,0 +1,261 @@
+import * as React from 'react';
+import {
+  BaseComponent,
+  anchorProperties,
+  autobind,
+  buttonProperties,
+  getId,
+  getNativeProps,
+  customizable,
+  nullRender
+} from '../../../Utilities';
+import { hiddenContentStyle, mergeStyles } from '../../../Styling';
+import { Icon } from '../../../Icon';
+import { IButtonBaseProps, IButtonBase, IButtonBaseStyleProps, IButtonBaseStyles } from './Button.base.types';
+import {
+  classNamesFunction
+} from '../../../Utilities';
+
+const getClassNames = classNamesFunction<IButtonBaseStyleProps, IButtonBaseStyles>();
+
+@customizable('ButtonBase', ['theme'])
+export class ButtonBase extends BaseComponent<IButtonBaseProps, {}> implements IButtonBase {
+
+  public static defaultProps = {
+  };
+
+  private _classNames: {[key in keyof IButtonBaseStyles]: string };
+  private _buttonElement: HTMLElement;
+  private _labelId: string;
+  private _descriptionId: string;
+  private _ariaDescriptionId: string;
+
+  constructor(props: IButtonBaseProps, rootClassName: string) {
+    super(props);
+
+    this._labelId = getId();
+    this._descriptionId = getId();
+    this._ariaDescriptionId = getId();
+  }
+
+  public render(): JSX.Element {
+    const {
+      theme,
+      className,
+      disabled,
+      href,
+      checked,
+      primary,
+      getStyles,
+      expanded,
+      onRenderIcon = this._onRenderIcon,
+      onRenderAriaDescription = this._onRenderAriaDescription,
+      onRenderChildren = this._onRenderChildren,
+      onRenderMenuIcon = this._onRenderMenuIcon,
+      onRenderPrefix,
+      onRenderSuffix
+    } = this.props;
+
+    // Anchor tag cannot be disabled hence in disabled state rendering
+    // anchor button as normal button
+    const renderAsAnchor: boolean = !disabled && !!href;
+
+    const nativeProps = this._getNativeProps(this.props, renderAsAnchor);
+
+    const {
+      ariaDescribedBy,
+      ariaLabelledBy,
+      ariaLabel
+    } = this._getAriaProps(
+        this.props,
+        getNativeProps(this.props, [renderAsAnchor ? 'a' : 'button']
+        ));
+
+    this._classNames = getClassNames(getStyles!, { theme: theme!, className, disabled, checked, expanded, primary });
+
+    let RootType = renderAsAnchor ? 'a' : 'button';
+    return (
+      <div className={ this._classNames.root }>
+        { onRenderPrefix && onRenderPrefix() }
+
+        <RootType
```

Comment:
Another question worth asking: 

Is it worth having the notion of onRenderPrefix/onRenderPostFix/onRenderMenuIcon/onRenderIcon instead of having clear render functions for each of the button components. For example, Split button would be <div><Button/><Button/></div>
button would just be <Button><Icon/><div>{text}</div></Button> etc...

Moving away from onRenderIcon is probably a good idea since we have support for "custom" icon rendering (such as SVGs) through the registerIcon API. Having undefined icon props would be the way to not render an icon rather than using a null render function

Reply:
lets talk more about this and what it'd look like